sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp

```diff
@@ -70,146 +76,668 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
+    operator std::uint64_t() const;
+    operator std::int64_t() const;
     operator float() const;
-    operator float();
     operator double() const;
-    operator double();
-    operator BinaryData() const;
-    operator BinaryData();
     explicit operator std::string() const;
-    explicit operator std::string();
-    operator Uuid();
     operator Uuid() const;
 
-    bool operator==(AmqpValue const& that) const;
-    AmqpValueType GetType() const;
-
     // List Operations.
-    static AmqpValue CreateList();
-    void SetListItemCount(uint32_t count);
-    uint32_t GetListItemCount() const;
-    void SetListItem(uint32_t index, AmqpValue item);
-    AmqpValue GetListItem(size_t index) const;
-    //    AMQPAmqpValue GetListItemInPlace(size_t index) const;
+    AmqpList AsList() const;
 
     // Map operations.
-    static AmqpValue CreateMap();
-    void SetMapValue(AmqpValue key, AmqpValue value);
-    AmqpValue GetMapValue(AmqpValue key) const;
-    std::pair<AmqpValue, AmqpValue> GetMapKeyAndValue(uint32_t index) const;
-    size_t GetMapValueCount() const;
+    AmqpMap AsMap() const;
 
     // Array operations - note that all array items must be of the same type.
-    static AmqpValue CreateArray();
-    void AddArrayItem(AmqpValue itemAmqpValue);
-    AmqpValue GetArrayItem(uint32_t index) const;
-    uint32_t GetArrayItemCount() const;
+    AmqpArray AsArray() const;
+
+    AmqpBinaryData AsBinary() const;
 
-    // Char
-    static AmqpValue CreateChar(uint32_t value);
-    uint32_t GetChar() const;
+    AmqpTimestamp AsTimestamp() const;
 
-    // Timestamps
-    static AmqpValue CreateTimestamp(std::chrono::milliseconds value);
-    std::chrono::milliseconds GetTimestamp() const;
+    // Get Char value.
+    std::uint32_t GetChar() const;
 
     // Symbols
-    static AmqpValue CreateSymbol(std::string const& value);
-    std::string GetSymbol() const;
+    AmqpSymbol AsSymbol() const;
 
     // Composite values - A composite value is functionally a list with a fixed size.
-    static AmqpValue CreateComposite(AmqpValue descriptor, uint32_t listSize);
-    void SetCompositeItem(uint32_t index, AmqpValue itemAmqpValue);
-    AmqpValue GetCompositeItem(uint32_t index);
-    //    AMQPAmqpValue GetCompositeItemInPlace(size_t index) const;
-    size_t GetCompositeItemCount() const;
+    AmqpComposite AsComposite() const;
 
-    static AmqpValue CreateDescribed(AmqpValue descriptor, AmqpValue value);
-    static AmqpValue CreateCompositeWithDescriptor(uint64_t descriptor);
+    AmqpDescribed AsDescribed() const;
 
-    // Descriptors
-    AmqpValue GetDescriptor() const;
-    AmqpValue GetDescribedValue() const;
+    friend std::ostream& operator<<(std::ostream& os, AmqpValue const& value);
 
-    // Headers.
-    bool IsHeaderTypeByDescriptor() const;
-    Header GetHeaderFromValue() const;
+  protected:
+    UniqueAmqpValueHandle m_value;
+  };
 
-    static AmqpValue CreateHeader(Header const& header);
+  namespace _detail {
```

Comment:
I'm not 100% sure if this should be in the `_detail` namespace or not. The public methods of this class are intended to be interactable from customers, because they are implementing the relevant STL functions that allow these types to be interacted with as if they were the corresponding STL type. But if they're in the `_detail` namespace, then they won't appear in the ApiView.

Reply:
We have the similar situation in `DateTime` - it is ok for customers to write `DateTime::duration`, but not ok to write `Azure::_detail::Clock::duration`. I think this is the similar case.
Another similar case is probably public typedeffing `_internal::UniqueHandle<PublicType>` - and it is ok for customers to use it, but not to write `_internal::...`